File: script/modules_collector.py
import os
import argparse
import yaml
import six
import glob
import re

# Collects module configs to modules.d

REFERENCE_CONFIG_RE = re.compile('.+\.reference\.yml')


def collect(beat_name, docs_branch):

    base_dir = "module"
    path = os.path.abspath("module")

    # TODO add module release status if beta or experimental
    header = """# Module: {module}
# Docs: https://www.elastic.co/guide/en/beats/{beat_name}/{docs_branch}/{beat_name}-module-{module}.html

"""

    # Create directory for module confs
    os.mkdir(os.path.abspath('modules.d'))

    # Iterate over all modules
    for module in sorted(os.listdir(base_dir)):

        module_confs = path + '/' + module + '/_meta/config*.yml'
        for module_conf in glob.glob(module_confs):

            # Ignore reference confs
            if REFERENCE_CONFIG_RE.match(module_conf):
                continue

            if os.path.isfile(module_conf) == False:
                continue

            module_file = header.format(module=module, beat_name=beat_name, docs_branch=docs_branch)
            disabled_config_filename = os.path.basename(module_conf).replace('config', module) + '.disabled'

            with open(module_conf) as f:
                module_file += f.read()

            # Write disabled module conf
            with open(os.path.abspath('modules.d/') + '/' + disabled_config_filename, 'w') as f:
                f.write(module_file)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Collects modules confs")
    parser.add_argument("--beat", help="Beat name")
    parser.add_argument("--docs_branch", help="Docs branch")

    args = parser.parse_args()
    beat_name = args.beat
    docs_branch = args.docs_branch

    collect(beat_name, docs_branch)

File: metricbeat/scripts/mage/modules_collector.go
// Licensed to Elasticsearch B.V. under one or more contributor
// license agreements. See the NOTICE file distributed with
// this work for additional information regarding copyright
// ownership. Elasticsearch B.V. licenses this file to you under
// the Apache License, Version 2.0 (the "License"); you may
// not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package mage

import (
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	devtools "github.com/elastic/beats/dev-tools/mage"
)

// CollectModules collects module configs to modules.d
func CollectModules() error {
	header := `# Module: %[1]s
# Docs: https://www.elastic.co/guide/en/beats/%[2]s/%[3]s/%[2]s-module-%[1]s.html

`
	r, err := regexp.Compile(`.+\.reference\.yml`)
	if err != nil {
		return err
	}

	beatName := devtools.BeatName
	docsBranch, err := devtools.BeatDocBranch()
	if err != nil {
		return err
	}

	path := devtools.OSSBeatDir("module")

	modules, err := ioutil.ReadDir("module")
	if err != nil {
		return err
	}

	if err = os.Mkdir("modules.d", os.ModeDir|os.ModePerm); err != nil {
		return err
	}

	modulesDDir := devtools.OSSBeatDir("modules.d")
	for _, module := range modules {
		moduleConfsGlob := filepath.Join(path, module.Name(), "_meta/config*.yml")
		moduleConfs, err := filepath.Glob(moduleConfsGlob)
		if err != nil {
			return err
		}

		for _, moduleConf := range moduleConfs {
			if r.MatchString(moduleConf) {
				continue
			}

			// skip directories
			if info, err := os.Stat(moduleConf); err != nil {
				return err
			} else if info.IsDir() {
				continue
			}

			moduleFile := fmt.Sprintf(header, module.Name(), beatName, docsBranch)
			disabledConfigFilename := strings.Replace(filepath.Base(moduleConf), "config", module.Name(), -1) + ".disabled"

			fileBytes, err := ioutil.ReadFile(moduleConf)
			if err != nil {
				return err
			}

			moduleFile += string(fileBytes)

			err = ioutil.WriteFile(filepath.Join(modulesDDir, disabledConfigFilename), []byte(moduleFile), 0644)
			if err != nil {
				return err
			}
		}
	}

	return nil
}
